anvil/observables.py

```diff
@@ -230,65 +341,61 @@ def autocorrelation_2pf(training_geometry, volume_averaged_2pf):
-    """
-    Nstates = len(G_series)
-    autocorr2 = torch.zeros(Nstates)
-    autocorr2[0] = G_series.pow(2).mean() - G_series.mean() ** 2
-    for t in range(1, Nstates):
-        term1 = torch.mean(G_series[:-t] * G_series[t:])
-        term2 = torch.mean(G_series[:-t]) * torch.mean(G_series[t:])
-        autocorr2[t] = term1 - term2
-    autocorr2 = autocorr2 / autocorr2[0]
-    """
+    n_states = int(G_series.size()[0])
+    tau_int_W = 0.5 + np.cumsum(autocorrelation[1:])
+    valid = np.where(tau_int_W > 0.5)[0]
+    tau_exp_W = np.ones(tau_int_W.size) * 0.00001  # to prevent domain error in log
 
-    integrated_autocorrelation = 0.5 + np.sum(autocorrelation[1:])
+    S = window_S  # read from runcard parameter
+    tau_exp_W[valid] = S / (
+        np.log((2 * tau_int_W[valid] + 1) / (2 * tau_int_W[valid] - 1))
+    )
+    W = np.arange(1, tau_int_W.size + 1)
+    g_W = np.exp(-W / tau_exp_W) - tau_exp_W / np.sqrt(W * n_states)
 
-    return autocorrelation, integrated_autocorrelation
+    W_opt = np.where(g_W < 0)[0][0]
+    w = 4 * W_opt  # where to cut the plot off
 
+    return autocorrelation[:w], tau_int_W[:w], tau_exp_W[:w], g_W[:w], W_opt
 
-@table
-def ising_observables_table(ising_energy, susceptibility, training_output):
-    res = [[ising_energy], [susceptibility]]
-    df = pd.DataFrame(
-        res, columns=[training_output.name], index=["Ising energy", "susceptibility"]
-    )
-    return df
-
-
-@figure
-def plot_zero_momentum_green_function(zero_momentum_green_function, training_geometry):
-    fig, ax = plt.subplots()
-    ax.plot(zero_momentum_green_function, "-r", label=f"L = {training_geometry.length}")
-    ax.set_yscale("log")
-    ax.set_ylabel(r"$\hat{G}(0, t)$")
-    ax.set_xlabel("t")
-    ax.set_title("Zero momentum Green function")
-    return fig
-
-
-@figure
-def plot_volume_averaged_2pf(volume_averaged_2pf):
-    fig, ax = plt.subplots()
-    ax.set_title("Volume-averaged two point function")
-    ax.set_ylabel(r"$G_V$")
-    ax.set_xlabel(r"$t$")
-    ax.plot(volume_averaged_2pf(0, 0), "-")
-    return fig
-
-
-@figure
-def plot_autocorrelation_2pf(autocorrelation_2pf):
-    autocorrelation, integrated_autocorrelation = autocorrelation_2pf
-    fig, ax = plt.subplots()
-    # ax.set_yscale("log")
-    ax.set_title("Autocorrelation of volume-averaged two point function")
-    ax.set_xlabel("$t$")
-    ax.set_ylabel("$\Gamma_{G_V}(t)$")
-    ax.plot(autocorrelation, "-")
-    x = 0.8 * (1 + len(autocorrelation))
-    y = 0.8
-    ax.text(
-        x, y, r"$\tau_{\rm int} = $ %.3g" % integrated_autocorrelation, fontsize="large"
-    )
-    return fig
+
+#######################
+###     Collect     ###
+#######################
+# I was hoping that doing this would mean we could loop over plotting parameters
+# without having to re-run the sampling, but it's not achieved this aim.
+_two_point_function = collect("two_point_function", ("training_context",))
```

Comment:
I think you should remove pretty much all the collects here, they are already either taking things like `sample_training_output` or `training_geometry` which already use `training_context` so collecting here doesn't help at all

Reply:
Ah sure. I recently added those in thinking they were the key to being able to loop over that `window_S` parameter in autocorrelation without re-running the sampling each time, but apparently they're not.